src/ansiblelint/transformer.py

```diff
@@ -0,0 +1,253 @@
+            file: [] for file in result.files
+        }
+
+        for match in self.matches:
+            try:
+                lintable = lintables[match.filename]
+            except KeyError:
+                # we shouldn't get here, but this is easy to recover from so do that.
+                lintable = Lintable(match.filename)
+                self.matches_per_file[lintable] = []
+            self.matches_per_file[lintable].append(match)
+
+        self._yaml: Optional[YAML] = None
+
+    @property
+    def yaml(self) -> YAML:
+        """Return a configured ``ruamel.yaml.YAML`` instance.
+
+        ``ruamel.yaml.YAML`` uses attributes to configure how it dumps yaml files.
+        Some of these settings can be confusing, so here are examples of how different
+        settings will affect the dumped yaml.
+
+        This example does not indent any sequences:
+
+        .. code:: python
+
+            yaml.explicit_start=True
+            yaml.map_indent=2
+            yaml.sequence_indent=2
+            yaml.sequence_dash_offset=0
+
+        .. code:: yaml
+
+            ---
+            - name: playbook
+              tasks:
+              - name: task
+
+        This example indents all sequences including the root-level:
+
+        .. code:: python
+
+            yaml.explicit_start=True
+            yaml.map_indent=2
+            yaml.sequence_indent=4
+            yaml.sequence_dash_offset=2
+            # yaml.Emitter defaults to ruamel.yaml.emitter.Emitter
+
+        .. code:: yaml
+
+            ---
+              - name: playbook
+                tasks:
+                  - name: task
+
+        This example indents all sequences except at the root-level:
+
+        .. code:: python
+
+            yaml.explicit_start=True
+            yaml.map_indent=2
+            yaml.sequence_indent=4
+            yaml.sequence_dash_offset=2
+            yaml.Emitter = FormattedEmitter  # custom Emitter prevents root-level indents
+
+        .. code:: yaml
+
+            ---
+            - name: playbook
+              tasks:
+                - name: task
+        """
+        if self._yaml is not None:
+            return self._yaml
+
+        # ruamel.yaml rt=round trip (preserves comments while allowing for modification)
+        yaml = YAML(typ="rt")
+
+        # NB: ruamel.yaml does not have typehints, so mypy complains about everything here.
+
+        # Ansible uses PyYAML which only supports YAML 1.1. ruamel.yaml defaults to 1.2.
+        # So, we have to make sure we dump yaml files using YAML 1.1.
+        # We can relax the version requirement once ansible uses a version of PyYAML
+        # that includes this PR: https://github.com/yaml/pyyaml/pull/555
+        yaml.version = (1, 1)  # type: ignore[assignment]
+        # Sadly, this means all YAML files will be prefixed with "%YAML 1.1\n" on dump
+        # We'll have to drop that using a transform so people don't yell too much.
```

Comment:
AFAIK, can't handle YAML 1.2 is not true because YAML 1.2 is a stricter set. Any tool that can load 1.1 can also load 1.2 the same.

1.1 gave users too much artistic freedom and one major fix in 1.2 is less freedom. This means that 1.1 documents can be converted to 1.2 without being worried about breaking them.

Reply:
For booleans, yes, it is a stricter set (which is great!). But numbers, especially octals, and some time parsing had their regexes modified in the 1.2 spec. So an octal in 1.2 (`0o755`) is considered a string in 1.1. And 1.2 does not recognize the 1.1 style octal (`0755`), and treats it as an int.

Reply:
I didn't say the issues were due to YAML 1.2, just that ruamel had generated YAML that Ansible (using libyaml) was unable to parse.

Reply:
the case of octals would have created different errors, but those would be seen after parsing (which did happen, but that was in addition to the other issue)

Reply:
in any case i was time constrained and just opted to loose comments and used pyyaml (a sure bet) vs trying to debug ruamel issues, so no idea how easy/hard it would be to handle them.

Reply:
K. Well, when someone runs across one of these issues (assuming this YAML 1.1 fix doesn't prevent them), I will debug it.

`ansible-lint` depends on comments to skip some rules, so people would get very angry if we started discarding their skip comments. So, discarding comments is not really an option for adding these features.

Reply:
not suggesting you do so, I'm just explaining the tradeoff I had to make during migration and the lack of investigation into the causes since I really did not have time to spare. I just gave a friendly warning that I had encountered these issues back then and that they merit investigation.

Reply:
Makes sense! Thank you!